### PR TITLE
feat(tokens): add design tokens architecture

### PR DESCRIPTION
**Pull request by BenOsodrac**
## What is the new behavior?

Following the Design Doc for the new Ionic Design System, an architecture for Design Tokens was added.

1. Added new `foundations` folder inside the `src`, to hold all the tokens related files.
2. On this file, a json with all the tokens for the new Design System was added. The format followed was the one suggested on the [W3C draft](https://design-tokens.github.io/community-group/format/).
3. Added a dev dependency for [Styles Dictionary](https://amzn.github.io/style-dictionary/#/), version 3.9.2.
4. Added a tokens.js script that will take care of transforming the json tokens into the desired output. For now, three files are being generated: ionic.vars.css (:root with all CSS Variables); ionic.vars.css (scss variables for each token) and ionic.utility.css (a new utility-class for each token)
5. Added the script `npm run build.tokens` to package.json, that will generate the three files mentioned above.

For now, all these changes bring no impact to the rest of the Framework, as these variables are not yet being consumed.

The `margin`, `padding` and `border-radius` were removed from the prohibited properties on lint, to prevent lint errors with the new utility-classes. This is very open to discussion/feedback if it's seen as not ok.

The `build` command now includes the `build.tokens` script, to make sure the files are always generated, in case someone forget to run the command, after changing the json file!

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/buildTokens.js
Comment (sean-perkins) [inline core/buildTokens.js:19]: Can we insert a comment block in the file contents that get generated that inform a user opening the file that it is auto-generated and they should not make changes in that file?

File: core/buildTokens.js
Comment (sean-perkins) [inline core/buildTokens.js:2]: Does `style-dictionary` have any provided types that we could use JSDoc annotations to import the type references for?

File: core/src/foundations/ionic.root.scss
Comment (sean-perkins) [inline core/src/foundations/ionic.root.scss:6]: Are we intentionally not prefixing these CSS variables to avoid conflicting with the existing `--ion` variables?

File: core/src/foundations/ionic.utility.scss
Comment (sean-perkins) [inline core/src/foundations/ionic.utility.scss]: ~ We should identify if prettier can format this output. This could be a technical debt ticket/not blocking this PR.

File: core/package.json
Comment (sean-perkins) [inline core/package.json]: Question for understanding, when should a developer run this command?

File: core/buildTokens.js
Comment (sean-perkins) [inline core/buildTokens.js:1]: Existing Ionic script files are lowercase and use kebab-case if multi-word. Can we rename this file and update the script command?

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js:19]: Yes, its in fact the default behaviour 😂, I will adjust that!

File: core/src/foundations/ionic.root.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.root.scss:6]: Yes, I was not sure at this moment what we would want to have going forward and I thought not using that a bit more safe. 

But that can be easily changed on the tokens script. Maybe it should be --ionic-, as this only affects the Ionic Design Systems and themes? But this is a decision that should be made on the scope of the rest of the Design Doc, as it also impacts the variables on the scope of each component for the Ionic Theme.

File: core/src/foundations/ionic.utility.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.utility.scss]: I already adjusted this on the script file and changed the `npm run tokens` script to also run `lint.sass.fix.` The generated file is different now.

Also, due to lint there were some errors that made me remove some prohibited properties (I added context on the PT description about this)

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js:1]: Ah, I actually had it named tokens.js and changed it to buildTokens.js as I saw the setupJest.js followed that approach. Will change it 👍

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js:2]: Hum, don't know, will try to check that!

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js]: Fixed!

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js]: Fixed!

File: core/src/foundations/ionic.root.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.root.scss]: Just added a prefix to all generated variables 👍

File: core/src/foundations/ionic.vars.scss
Comment (sean-perkins) [inline core/src/foundations/ionic.vars.scss:6]: What's the purpose of these sass variables? On a closer inspection they will be statically set based on the `_design-tokens.json`. If we reference these variables in our components, developers will not be able to customize the value.

I think these either need to be set to the CSS variable, with the default fallback to the static value or we don't generate them at all and reference the CSS variables directly.

File: core/src/foundations/ionic.root.scss
Comment (sean-perkins) [inline core/src/foundations/ionic.root.scss:7]: ~ Non-blocking, we should see if we can format this or generate the output to match our indentation for CSS variables.

File: core/src/foundations/ionic.utility.scss
Comment (sean-perkins) [inline core/src/foundations/ionic.utility.scss:1]: ~ Non-blocking, same feedback here regarding formatting. If it isn't simple, we can come back and have prettier format these files.

File: core/src/foundations/ionic.vars.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.vars.scss:6]: The purpose was to follow a similar approach as we already have on the md/ios, where only scss variables are used internally and no global :root is defined by default. The would be used internally at component level and to set the values on the global Ionic the defaults, as suggested here: https://github.com/BenOsodrac/ionic-framework/blob/new-ionic-ds/core/src/themes/ionic/ionic.theme.default.scss

But I'm ok with using only the native CSS Variables, its in fact more inline on how I'm used to do things 👍

Comment (sean-perkins) [reply core/src/foundations/ionic.vars.scss:6]: I think it makes sense to have scss variables to use internally, but I think they need to be like `$background-color-step` variables: https://github.com/BenOsodrac/ionic-framework/blob/f3e01bb0057f6c917d582efdf4208829cc6eb14f/core/src/themes/ionic/ionic.theme.default.scss#L106.

With the current state, if I had this:
```
/* button.ionic.scss */
:host {
  color: $ionic-color-primary-10;
}
```

When the stylesheet is compiled the generated output for the developer consuming would be:
```
:host {
  color: #f7faff;
}
```

If a developer tries to customize the color palette in their application:
```
:root {
  --ionic-color-primary-10: orange;
}
```

The button stylesheet will still be `#f7faff`. This isn't desired. 

However if we have:
```
$ionic-color-primary-10: var(--ionic-color-primary-10);
```

Then a developer consuming the components can customize the color used in the component globally.

File: core/buildTokens.js
Comment (BenOsodrac) [reply core/buildTokens.js]: The have ts types: https://github.com/amzn/style-dictionary/tree/main/types, would that be useful for JSDoc?

File: core/src/foundations/ionic.vars.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.vars.scss]: Your PR is merged 👍, lets go 🚀

**Review by sean-perkins — APPROVED** [review]
Great work on this 👍

File: core/package.json
Comment (liamdebeasi) [inline core/package.json:104]: This script should probably be run on build. My main concern here is what happens if someone modifies the JSON file but forgets to run this script? We'd potentially ship with outdated styles.

I'd recommend doing something like this:

1. Update the `build` script above to call `npm run tokens` (You could also rename this to `build.tokens` if you wanted naming consistency). This ensures that the latest tokens are generated on `npm run build`. Note that this should be run **before** `npm run build.css` since the tokens script outputs scss.
2. In the [`build-core` workflow](https://github.com/ionic-team/ionic-framework/blob/b5cb828861c568919a8a8f1090677f37dc1ce424/.github/workflows/actions/build-core/action.yml#L34)  and the [`build-core-stencil-prerelease` workflow](https://github.com/ionic-team/ionic-framework/blob/b5cb828861c568919a8a8f1090677f37dc1ce424/.github/workflows/actions/build-core-stencil-prerelease/action.yml#L32) make sure the `core/src/foundations` directory is cached.

The benefit of this is if anyone forgets to build the latest scss file the [`test-core-clean-build`](https://github.com/ionic-team/ionic-framework/blob/main/.github/workflows/actions/test-core-clean-build/action.yml) will fail in CI.

File: core/src/foundations/ionic.vars.scss
Comment (liamdebeasi) [inline core/src/foundations/ionic.vars.scss:6]: If these variables are guaranteed to always be defined due to `ionic.root.scss` why we do we need to repeat the fallback values here?

File: core/tokens.js
Comment (liamdebeasi) [inline core/tokens.js:2]: Can we add a sentence or two about why we are doing this? For example, it would be good to document why we use `style-dictionary` as opposed to just hardcoding these values directly.

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
Overall looks good! I have some questions surrounding long term maintenance of this.

File: core/package.json
Comment (BenOsodrac) [reply core/package.json]: Yes, that makes sense, already did the change!

Comment (BenOsodrac) [reply core/package.json]: Hum thank you for this feedback, I think it makes total sense 👍 

1. Already updated the script name and move it to the build command.
2. Added the `core/src/foundations` to the path attribute of those actions. I don't have much context about this, is there anything more I should do?

File: core/tokens.js
Comment (BenOsodrac) [reply core/tokens.js]: Done!

File: core/src/foundations/ionic.root.scss
Comment (BenOsodrac) [reply core/src/foundations/ionic.root.scss:7]: Yes, we did! 

In summary, as we are introducing a new Design System, we wanted to avoid confusion and conflicts with the current one. 

As the new theme is also called Ionic, we thought the `--ionic` prefix makes total sense for these new tokens. 

This decision is also present on the [Decision Log](https://outsystemsrd.atlassian.net/wiki/spaces/RDMBVS/pages/4057366529/Native+Look+Feel+-+Decisions+Log) page 👍

File: core/src/foundations/ionic.vars.scss
Comment (sean-perkins) [reply core/src/foundations/ionic.vars.scss:6]: This is to ensure that the variables always have a value assigned to them. Developers should be importing a global stylesheet that includes the content of `ionic.root.scss` or assigning their own values to these variables. 

This approach is most helpful in cases where the variable assignment may fail or not be initially available at the time of render. 

It is better to have a fallback experience/value vs. a completely broken experience. 

If we decided to remove this in the future, it is pretty trivial to update the script that generates that format.

File: core/src/foundations/ionic.root.scss
Comment (liamdebeasi) [reply core/src/foundations/ionic.root.scss:7]: sounds good, thanks!

File: core/package.json
Comment (liamdebeasi) [reply core/package.json]: Perfect, thanks!

Comment (liamdebeasi) [reply core/package.json]: Looks great! I tested this out on a separate branch, and CI correctly fails if you change the source json file without re-running the build: https://github.com/ionic-team/ionic-framework/actions/runs/8560100688/job/23458371826?pr=29276

File: .github/workflows/actions/build-core-stencil-prerelease/action.yml
Comment (liamdebeasi) [inline .github/workflows/actions/build-core-stencil-prerelease/action.yml:9]: Can we revert the quotation marks change?

File: .github/workflows/actions/build-core/action.yml
Comment (liamdebeasi) [inline .github/workflows/actions/build-core/action.yml:9]: Can we revert the quotation marks change?

**Review by liamdebeasi — APPROVED** [review]
Good to go on my end once the below changes are made. Great work!

File: .github/workflows/actions/build-core-stencil-prerelease/action.yml
Comment (BenOsodrac) [reply .github/workflows/actions/build-core-stencil-prerelease/action.yml]: Hum some stupid default config from Prettier, already changed it!

File: .github/workflows/actions/build-core/action.yml
Comment (BenOsodrac) [reply .github/workflows/actions/build-core/action.yml]: Done!

**Review by sean-perkins — APPROVED** [review]

**Pull request merged**
